packages/discord.js/src/structures/Message.js

```diff
@@ -439,8 +351,212 @@ class Message extends Base {
+    return typeof this.content === 'string' ? Util.cleanContent(this.data.content, this.channel) : null;
+  }
+
+  /**
+   * Whether the message is editable by the client user
+   * @type {boolean}
+   * @readonly
+   */
+  get editable() {
+    const precheck = Boolean(this.author.id === this.client.user.id && (!this.guild || this.channel?.viewable));
+    // Regardless of permissions thread messages cannot be edited if the thread is locked.
+    if (this.channel?.isThread()) {
+      return precheck && !this.channel.locked;
+    }
+    return precheck;
+  }
+
+  /**
+   * Whether the message is deletable by the client user
+   * @type {boolean}
+   * @readonly
+   */
+  get deletable() {
+    if (!this.guild) {
+      return this.data.author.id === this.client.user.id;
+    }
+    // DMChannel does not have viewable property, so check viewable after proved that message is on a guild.
+    if (!this.channel?.viewable) {
+      return false;
+    }
+
+    const permissions = this.channel?.permissionsFor(this.client.user);
+    if (!permissions) return false;
+    // This flag allows deleting even if timed out
+    if (permissions.has(PermissionFlagsBits.Administrator, false)) return true;
+
+    return Boolean(
+      this.author.id === this.client.user.id ||
+        (permissions.has(PermissionFlagsBits.ManageMessages, false) &&
+          this.guild.me.communicationDisabledUntilTimestamp < Date.now()),
+    );
+  }
+
+  /**
+   * Whether the message is pinnable by the client user
+   * @type {boolean}
+   * @readonly
+   */
+  get pinnable() {
+    const { channel } = this;
+    return Boolean(
+      !this.system &&
+        (!this.guild ||
+          (channel?.viewable &&
+            channel?.permissionsFor(this.client.user)?.has(PermissionFlagsBits.ManageMessages, false))),
+    );
+  }
+
+  /**
+   * Whether the message is crosspostable by the client user
+   * @type {boolean}
+   * @readonly
+   */
+  get crosspostable() {
+    const bitfield =
+      PermissionFlagsBits.SendMessages |
+      (this.author.id === this.client.user.id ? PermissionsBitField.defaultBit : PermissionFlagsBits.ManageMessages);
+    const { channel } = this;
+    return Boolean(
+      channel?.type === ChannelType.GuildNews &&
+        !this.flags.has(MessageFlags.Crossposted) &&
+        this.type === MessageType.Default &&
+        channel.viewable &&
+        channel.permissionsFor(this.client.user)?.has(bitfield, false),
+    );
+  }
+
+  _patch(data) {
+    const {
+      embeds,
+      components,
+      mentions,
+      mention_roles,
+      mention_channels,
+      mention_everyone,
+      attachments,
+      stickers,
+      sticker_items,
+      thread,
+      member,
+      reactions,
+      ...initData
+    } = data;
+
+    /**
+     * The raw API data for this message
+     * @type {APIMessage}
+     * @readonly
+     */
+    this.data = { ...initData, ...this.data };
```

Comment:
If `APIMessage` is supposed to be the type from -types then it's not correct, as it's missing `attachments`, `components`, etc.

Reply:
Yeah, I meant `this.data` doesn't contain the things that I listed. So APIMessage is not the correct type for `this.data`

Reply:
Oh well I'm not really sure how to `Omit` in jsdoc (without listing every individual prop) so I left in the base type.